### PR TITLE
Spark 3.3: Handle Spark Row more generically

### PR DESCRIPTION
**Pull request by flyrain**
cc @aokolnychyi @szehon-ho @RussellSpitzer

### CODE REVIEW

**Comment by aokolnychyi** [comment]
Will take a look today.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java [hunk 1]
```diff
@@ -83,12 +85,16 @@ private ChangelogIterator(
    * @param changeTypeIndex the index of the change type column
    * @param identifierFieldIdx the indices of the identifier columns, which determine if rows are
    *     the same
+   * @param columnSize the number of columns in the table
    * @return a new {@link ChangelogIterator} instance concatenated with the null-removal iterator
    */
-  public static Iterator<Row> iterator(
-      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> identifierFieldIdx) {
+  public static Iterator<Row> create(
```
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java:91]: What about making the signature as follows?

```
public static Iterator<Row> create(
    Iterator<Row> rowIterator,
    StructType rowType,
    List<String> identifierFields) {

  // compute indices for the change and identifier columns here
}
```

I feel like we accept some low-level details in this method. Instead, I'd accept an iterator, schema, and identifier column names to move the complexity from the procedure here.

Comment (aokolnychyi) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java:91]: I believe you should already have this code in the procedure. I'd move it here.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java [hunk 2]
```diff
@@ -125,18 +127,32 @@ public Row next() {
           currentRow = null;
           cachedRow = null;
         } else {
-          Row[] rows = createUpdateChangelog((GenericRowWithSchema) currentRow, nextRow);
```
Comment (aokolnychyi) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java:128]: Looks better!

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java [hunk 3]
```diff
@@ -147,19 +163,6 @@ private int[] generateIndicesForIdentifySameRow(Row row) {
...
-    insertedRow.update(changeTypeIndex, UPDATE_AFTER);
-
-    return new Row[] {
-      RowFactory.create(deletedRow.values()), RowFactory.create(insertedRow.values())
-    };
-  }
-
   private boolean isCarryoverRecord(Row currentRow, Row nextRow) {
     for (int idx : indicesForIdentifySameRow) {
       if (!isColumnSame(currentRow, nextRow, idx)) {
```
Comment (clesaec) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java:168]: My 2 cents, it seems that modify this method as :
```java
  private boolean isCarryoverRecord(Row currentRow, Row nextRow) {
    for (int idx = 0; idx < currentRow.length(); idx++) {
      if (idx != this.changeTypeIndex) {
        if (!isColumnSame(currentRow, nextRow, idx)) {
          return false;
        }
      }
    }

    return true;
  }
```
allows to remove indicesForIdentifySameRow array

Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java:168]: To precompute would be faster than to check within the loop.  More details are in https://github.com/apache/iceberg/pull/6344/files/e813670402459901ed4f2b36cb24349145b705ac#r1060949312.

**Comment by flyrain** [comment]
Thanks @aokolnychyi and @clesaec for the review. Resolved the comments in a new commit. Please take a look.

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java (hunk 1 above)
Comment (flyrain) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java:94]: That's a good idea. Fixed it in the new commit.

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks, @flyrain!

**Pull request merged**